mediatek: filogic: add Adtran SmartRG Mount Stuart series

dangowrt opened the pull request:
This series adds basic support for the Adtran SmartRG [SDG 8700 Series](https://www.adtran.com/en/products-and-services/residential-solutions/mesh-wi-fi-gateways-and-satellites/service-delivery-gateways/sdg-8700-series).

Common specifications:
 * Mediatek MT7988A (4x Cortex-A73, up to 1.8 GHz clock speed)
 * 8 GiB eMMC
 * 2 GiB DDR4 RAM
 * 1x 10000M/1000M/100M, 3x 1000M/100M/10M LAN ports
 * MT7996 Tri-band (2.4GHz, 5GHz, 6GHz) 4T4R 802.11be Wi-Fi
 * Airoha AG3352 GPS
 * Renesas DA14531MOD Bluetooth
 * 2 buttons (Reset, Mesh/WPS)
 * uC-controlled color LED via I2C
 * 2x LED for each 1G port, 3x LED for each 10G port
 * USB 3.0 type A port
 * 3.3V-level 115200baud UART console via 4-pin Dupont connector exposed at the bottom of the device
 * USB-C PD power input

SDG-8733: 1x 10000M/1000M/100M WAN port
SDG-8734: 1x USXGMII/10GBase-R/5GBase-R/2500Base-X/1000Base-X/SGMII SFP+

Both models are also available in versions including 2x FXS POTS interfaces for connecting analog phones. Those interfaces are not supported by OpenWrt.

dangowrt committed:
mediatek: filogic: add Adtran SmartRG Mount Stuart series

Common specifications:
 * Mediatek MT7988A (4x Cortex-A73, up to 1.8 GHz clock speed)
 * 8 GiB eMMC
 * 2 GiB DDR4 RAM
 * 1x 10000M/1000M/100M + 3x 1000M/100M/10M LAN ports
 * MT7996 Tri-band (2.4G, 5G, 6G) 4T4R 802.11be Wi-Fi
 * Airoha AG3352 GPS
 * Renesas DA14531MOD Bluetooth
 * 2 buttons (Reset, Mesh/WPS)
 * uC-controlled RGB LED via I2C
 * 2x LED for each 1G port, 3x LED for each 10G port
 * USB 3.0 type A port
 * 3.3V-level 115200 baud UART console via 4-pin Dupont connector
   exposed at the bottom of the device
 * USB-C PD power input

SDG-8733: 1x 10000M/1000M/100M WAN port
SDG-8734: 1x USXGMII/10GBase-R/5GBase-R/2500Base-X/1000Base-X/SGMII SFP+

Both models are also available in versions including 2x FXS POTS interfaces
for analog phones. Those interfaces are not supported by OpenWrt.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
mediatek: fix sysupgrade config backup on Adtran SmartRG devices

Add missing call to emmc_copy_config which either writes the sysupgrade
tar.gz backup file or clears the existing rootfs_data overlay.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

mrkiko commented:
So there is no Wi-Fi hardware?
Did I miss the installation instructions?
Great work.

dangowrt commented:
Thank you for reviewing!

> So there is no Wi-Fi hardware?

No, I just forgot to mention it. It's MT7996 tri-band 802.11be, added to the PR and commit description now.

> Did I miss the installation instructions?

No, you didn't miss them, but they are missing. I will add instructions.

cmonroe commented:
I tried your changes on an 8734 today and everything seems to work great.

My only recommendation would be to change the 1G PHY LEDs to illuminate solid on link up. Something like this:
```
system.led_lan_2_green.mode='link_1000'
system.led_lan_2_amber.mode='link_100 link_10'
system.led_lan_3_green.mode='link_1000'
system.led_lan_3_amber.mode='link_100 link_10'
system.led_lan_4_green.mode='link_1000'
system.led_lan_4_amber.mode='link_100 link_10'
```

I'm not sure how  much it matters but for the front/system LED we normally flash green during kernel init then switch to solid white. Flashing/pulsing red is used during upgrade. Perhaps something like this (aliases for these LEDs would need to be added too):
```
led-boot = &sys_green;
led-failsafe = &sys_blue;
led-running = &sys_white;
led-upgrade = &sys_red;
```
I'm more than happy to do a PR afterwards with this too.

Thanks Daniel!

cmonroe commented:
One thing I forgot to ask: is there any type of LED trigger for a SFP? These units have a small light pipe underneath the SFP cage with 2 SoC GPIO controlled LEDs (green and red). The intention was to convey 2 states:
1. SFP module detected (via mod-def0) but no link - illuminate red
2. SFP module detected and link detected - illuminate green

Neither would be illuminated if no SFP module is inserted.

I've not  really found a good way to do this so I just left the SFP LED unused for now.

cmonroe commented:
Sorry for the spam but I thought of one more thing. I didn't find it in a quick search of the codebase but if there's a standard way to define /etc/fw_env.config for a given target we could populate it as follows to provide access to u-boot-env:
```
root@OpenWrt:/# cat /etc/fw_env.config 
/dev/mmcblk0p1  0x0000  0x8000
```

dangowrt commented:
> One thing I forgot to ask: is there any type of LED trigger for a SFP?

Unfortunately Linux doesn't have that feature yet but I do agree that we need it. Actually LED triggers and SFP modules kinda became enemies in the Linux code before and recently caused some ugly involuntary fallout when support for PHY LEDs was added... As a stop-gap solution (to prevent kernel Oops on removal of some RJ-45 SFP modules) I submitted [a patch](https://github.com/torvalds/linux/commit/d98ce1f083cca895982f3e88b8e8c372288e3e23) which prevents the ancient single-speed triggers from being registered if the PHY is found on a SFP module (ie. on a RJ-45 SFP). Revisiting this construction site has been on my agenda for a while...

jcdutton commented:
Do you have any details for the device?
Is there a pinout for the UART port? 
How do you select different boot devices, and which can you choose from?

dangowrt commented:
> Is there a pinout for the UART port?

It's a 2.54mm 2x2-pin dupont connector on the bottom of the device, protected with a rubber cover.
```
4 . . 3
1 . . 2
```
1 RX
2 GND
3 3.3V
4 TX



> How do you select different boot devices, and which can you choose from?

This series of devices all got an eMMC as the only available boot device. So it always boots from eMMC, unless you chose to boot via UART using [mtk_uartboot](https://github.com/981213/mtk_uartboot) for recovery.

jcdutton commented:
I have found some more information.
With the UART connected:
When message "autoboot in 2 seconds" appears press the 's' then 't' key to reach the uboot prompt.

You can then boot from usb by using standard uboot commands like "usb start" etc.  or use "bootmenu" etc.

In general uboot escape to the prompts on various devices I have seen are:
1) Press any key to escape
2) Press " " (space) to escape.
3) Press CTRL-C to escape.
4) Type the letters "stop" to escape.
5) Press ESC key to escape.
The adtran appears on only need the "st" of stop.

dangowrt committed:
generic: 6.6: backports and add LED support for Aquantia PHYs

Backport patch adding support for the AQR114C PHY and add support for
PHY LEDs and polarity setting of Aquantia 3rd and 4th generation PHYs.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

Ansuel commented:
Oh wow was waiting for a device that finally had the aquantia LEDs attached!

Ansuel commented on target/linux/generic/pending-6.6/743-net-phy-aquantia-add-support-for-PHY-LEDs.patch:
```diff
@@ -0,0 +1,515 @@
++	}
++
++	if (active_low)
++		priv->led_polarities |= BIT(index);
```
Mhhh why do we need to track led_polarities state? From what I can see (in my documentation I don't have regs for led polarity... maybe you can help me with that) each LED can be independently set for the led polarity to active-low hence we can just apply the registry instead of having the reg to a separate function.

dangowrt replied:
We could apply again from DT after each reset instead. The idea was to only parse DT in the probe function once.

Ansuel replied:
Mhh but that is already done here

https://elixir.bootlin.com/linux/latest/source/drivers/net/phy/phy_device.c#L3289

when the OF led is probed the led_polarity_set is called for each LED.

dangowrt replied:
Yes, but [`of_phy_led()`](https://elixir.bootlin.com/linux/latest/C/ident/of_phy_led) is only called once on attach and not again on [`phy_init_hw()`](https://elixir.bootlin.com/linux/latest/C/ident/phy_init_hw) which would be required for the setting to survive reset. It would of course be possible for phylink to take care of that and I also wasn't sure whether it should be done there or in the phy driver, so I asked on netdev list. Andrew told me that keeping the value in `priv` and having the phy driver take of memorizing it would be the preferred solution.

Ansuel commented on target/linux/generic/pending-6.6/743-net-phy-aquantia-add-support-for-PHY-LEDs.patch:
```diff
@@ -0,0 +1,515 @@
+ 		   fw_major, fw_minor, build_id, prov_id);
+ }
+ 
++static int aqr107_leds_init(struct phy_device *phydev)
++{
++	int index, ret;
++
++	if (!IS_ENABLED(CONFIG_OF_MDIO))
++		return 0;
++
++	for (index = 0; index < AQR_NUM_LEDS; ++index) {
++		ret = aqr_phy_led_polarity_apply(phydev, index);
++		if (ret)
++			return ret;
++	}
++
++	return 0;
```
In theory this can be dropped. LED polarity ops are already called by generic PHY LEDs.

dangowrt replied:
Yes, but unfortunately they are not applied again after reset. I've inquired netdev list on the preferred solution for that and Andrew replied:
https://lore.kernel.org/netdev/bb146832-30fc-4c76-a083-51a1bc087e61@lunn.ch/

Ansuel replied:
@dangowrt can you test this? 

https://gist.github.com/Ansuel/3f6b1ce80df26e58a4b1d6042bd8d056

Ansuel replied:
@dangowrt nice... lets see what they say upstream maybe there is chance. Really hate watching the same reg applied 2 times and have all additional code for nothing

dangowrt replied:
@Ansuel are you going to send the patch moving LED settings from probe to init to netdev? net-next re-opens on Monday, so If you like you could also already include the commit adding support for the Aquantia PHY LEDs.
As everything seems to be working nicely now, I would merge this PR first of all.

Ansuel replied:
Yep maybe I will email Andrew just to make things quicker and see if he is maybe ok with it.

Ansuel commented:
@dangowrt aside from the 2 small comments the led implementation looks good... also i can see the thing merged pretty quick in net-next. I you want I can take care of it and submit in your behalf (keeping the From and SoB)

cmonroe commented on target/linux/mediatek/dts/mt7988a-smartrg-mt-stuart.dtsi:
```diff
@@ -0,0 +1,684 @@
+
+		reset-gpios = <&pio 62 GPIO_ACTIVE_LOW>;
+		reset-assert-us = <100000>;
+		reset-deassert-us = <1000000>;
```
I think I got these values from an older version of the vendor SDK.. they do seem excessive.

The current vendor SDK uses assert = 100000; deassert-us = 221000.

Is there a good way to determine better/ideal values for these? If it matters, the firmware is loaded automatically from an external SPI flash (rather than MDIO download) so maybe that's why the deassert time is a bit longer?

Ansuel replied:
Oh god... those reset times are to wait for the PHY to finish loading the fw? :/ Did anyone test to reduce them to the spec Aquantia suggest? 

should be 100000 us for assert and 20000 deassert (100ms RESET 20ms RESET_OUT)

cmonroe replied:
That's my guess, I don't know for sure..  at least for designs which have a SPI flash attach like this one.

I tried 100ms assert/20ms deassert and the PHYs actually could link and pass traffic but they are missing any config (e.g. LED config) and the 2nd PHY (lan1/MDIO address 8) is identified as a Generic Clause 45 PHY instead of Aquantia AQR113C. I tried a few more values and it seems these things sure take their time. In all cases I used 100ms assert.

20ms deassert = fail (phy incorrectly identified)
200ms deassert = fail (phy incorrectly identified)
300ms deassert = fail (phy incorrectly identified)
400ms deassert = book OK, some LED config lost after network restart
500ms deassert = book OK, some LED config lost after network restart
600ms deassert = book OK, some LED config lost after network restart
800ms deassert = book OK, some LED config lost after network restart
1000ms deassert = OK

All that to say: it does look like these things really need close to a second between deassert and start of config at least when getting firmware from an attached SPI flash.

dangowrt commented:
@jfranzlx You can ask Ben Greear of Candelatech, I know they are at least selling the SDG-8733, also single devices, and also to end customers. You may be lucky to obtain the SDG-8734 through them as well.

https://www.candelatech.com/adtran_sdg_8733_dev.php
